framework/src/base/MooseApp.C

```diff
@@ -347,37 +352,61 @@ MooseApp::MooseApp(InputParameters parameters)
 #ifdef HAVE_GPERFTOOLS
   if (isUltimateMaster())
   {
-    if (std::getenv("MOOSE_PROFILE_BASE") && std::getenv("MOOSE_HEAP_BASE"))
-      mooseError("Can not do CPU and heap profiling together");
-
+    static std::string profile_file;
     // For CPU profiling, users need to have envirement 'MOOSE_PROFILE_BASE'
     if (std::getenv("MOOSE_PROFILE_BASE"))
     {
-      static std::string profile_file =
-          std::getenv("MOOSE_PROFILE_BASE") + std::to_string(_comm->rank()) + ".prof";
-
       _cpu_profiling = true;
-
-      auto name = MooseUtils::splitFileName(profile_file);
-      if (!name.first.empty() && processor_id() == 0)
-        MooseUtils::makedirs(name.first.c_str());
-
-      if (!ProfilerStart(profile_file.c_str()))
-        mooseError("CPU profiler is not started properly");
+      profile_file = std::getenv("MOOSE_PROFILE_BASE") + std::to_string(_comm->rank()) + ".prof";
     }
 
     // For Heap profiling, users need to have 'MOOSE_HEAP_BASE'
     if (std::getenv("MOOSE_HEAP_BASE"))
     {
-      static std::string profile_file =
-          std::getenv("MOOSE_HEAP_BASE") + std::to_string(_comm->rank());
-
       _heap_profiling = true;
+      profile_file = std::getenv("MOOSE_HEAP_BASE") + std::to_string(_comm->rank()) + ".prof";
+    }
+
+    if (_cpu_profiling && _heap_profiling)
+      mooseError("Can not do CPU and heap profiling together");
 
+    if (_cpu_profiling || _heap_profiling)
+    {
       auto name = MooseUtils::splitFileName(profile_file);
       if (!name.first.empty() && processor_id() == 0)
         MooseUtils::makedirs(name.first.c_str());
+      _comm->barrier();
```

Comment:
We possibly do  not need this `barrier`? What is happening if we remove this `barrier`?

Reply:
we have to otherwise the other processors will run pass this point and access a directory that has not been created by head processor. This is confirmed by my runs on Sawtooth.